docs/enhancements/auto-return-agent-to-infra-env.md

```diff
@@ -0,0 +1,140 @@
+the cluster was installed to run an agent on the host. This agent will handle
+downloading the required artifacts and rebooting the host into the discovery
+process.
+
+### Timeline
+
+1. User triggers an action which will prepare a host to be unbound
+  * Assisted service will create a MachineConfig to run our agent on the host as
+    a systemd service.
+  * The agent will download the InfraEnv's kernel and initrd (including
+    discovery ignition) to the host's `/boot` partition.
+  * The agent will create a GRUB boot entry to boot into the discovery kernel
+    and initrd.
+2. User (or other automation) removes the node from the cluster
+  * Drain existing workloads and delete the node
+3. User unbinds the host from the cluster using the assisted-service API
+  * This triggers the agent to reboot the host into the discovery image
+4. Host is registered as `known-unbound` in the InfraEnv and is ready to be used
+   in new clusters
+
+### User Stories
+
+#### Story 1
+
+As a cluster creator I want to use the assisted service to repurpose hosts 
+from existing clusters without requiring access to the host management 
+infrastructure (BMC, hypervisor, etc).
+
+### Implementation Details/Notes/Constraints [optional]
+
+1. A way to query for URLs to the iPXE boot artifacts over the REST API should be
+implemented. Currently these URLs are only exposed through the iPXE script and
+in the InfraEnv CR, but the agent running on a node will need access to these
+artifacts.
+  * This can be added to the existing payload returned when querying an InfraEnv
+  * Alternatively this would need a new API endpoint
+    * `/v2/infra-envs/{infra_env_id}/boot-artifacts`?
+2. The agent will likely require two changes/additions
+  * A step to query the boot artifacts and write the boot entry
+  * A step to reboot the host into the new boot entry
+3. The service will need a new API to prepare a cluster for hosts to be reclaimed
+  * This needs to be separate from unbind so that the user has time to cleanly
+    remove a host from the cluster
+  * `/v2/clusters/{cluster_id}/prepare-unbind`?
+
+#### Assumptions
+
+This approach only works if the host is currently part of a working cluster
+which the assisted service can access using the kubeconfig generated during
+install.
+
+There will also need to be enough space on the `/boot` partition to download the
+initrd and kernel, but the agent can check for this during the preparation step.
+
+### Risks and Mitigations
+
+It's possible that something could go wrong during the boot process. A user 
+that doesn't have access to the host infrastructure wouldn't be able to debug 
+or solve this type of problem. In this case the user is no worse off than they
+currently would be and will have to boot the discovery ISO manually.
+
+## Design Details [optional]
+
+### Open Questions
+
+1. Will users want to reclaim individual hosts in an existing cluster or will
```

Comment:
I think the agent will have different logic when removing a single node (cordon, wait for drain and reboot) vs decommissioning the cluster - it shouldn't care about the cordon but should wait with the reboot until all nodes are running the agent (so no one is left behind)

Reply:
Yeah, that's why this was split into two steps.
1. Run the agent on the hosts
** User or other automation ensures the cluster is ready for hosts to be removed here **
2. Reboot into discovery

I think I described this in the `Timeline` section of `Proposal`

Reply:
With the current proposal I don't think there's really any way for us to know the state of the other nodes in the cluster from the service.

It would just rely on the user doing everything they could before actually telling assisted-service to reclaim the node which would then directly cause it to reboot.

I think it would be enough to document this.